docs/design/kademlia-removal.md

```diff
@@ -0,0 +1,84 @@
+storage nodes, network refreshes, and kademlia removal given the satellite opt-in 
+capability for storage nodes. 
+
+## Background
+User research on how Storage Node Operators would connect to satellites resulted in a 
+simple solution called Opt-in SNO select. Satellites wait for storage nodes to work 
+with them, and storage nodes manually select the satellites they want to work with. 
+The initial implementation of this allows SNOs to update their trusted satellite list 
+in a hardcoded configuration file, but future improvements will enable users to manage 
+this list through a web console. Due to the nature of this change, the Storj network 
+at this point in time no longer requires the Kademlia DHT or related entities. 
+We will replace this protocol with direct communication between satellites and 
+storage nodes, and keep the network fresh without kademlia node discovery and random 
+lookups.
+
+## Design
+
+1. Nodes reach out to satellites that they want to work with
+    
+    - At the time of writing, a storage node operator can input a list of trusted satellite 
+    IDs and addresses into their configuration file on setup. There are several tardigrade-level 
+    satellites already listed as a preset. This list is imported into a Go map data structure 
+    within the “trust” package.
+    - Eventually during the design and implementation of trusted satellite management, 
+    this configured list may be imported into a on-disk data store, eg SQL table. Users will 
+    be able to modify this list through the storage node operator console. However, this is a 
+    separate discussion from this design document. This design is satisfied with a static trusted 
+    satellites list.
+    - Nodes will be able to communicate with satellites through the “transport” and “trust” packages 
+    directly rather than using kademlia to traverse the network to find the address of a given ID.
+
+2. Network refreshes at a regular interval
+    - Nodes will keep themselves up to date in the network by pinging all the satellites in their 
+    trusted list every hour. 
+    - Satellites will ping the nodes back
+        - If is it successful, the satellite will insert or update the node in the overlay cache and 
+        notify the node of success
+        - If not, it does not proceed with updating the overlay cache, and the node receives an error message
+    - This will use grpc for now, but future development may use a non-ssl connection from node to satellite 
+    to initiate the refresh connection
+    - Update the overlay cache refresh method to communicate with nodes directly, without relying on 
+    kademlia lookups. Iterate through cache and dial each node.
+
+3. Disintegrate Kademlia from the network, storj sim and testplanet setups
+    - Remove the “discovery” package
+    - Remove the bootstrap node - work with Ops
+    - Work with QA to make sure storage nodes don’t crash on errors related to the elimination of Kademlia 
```

Comment:
a storage node that upgrades straight to C will just appear offline to A. that's fine, right? I think we can skip B.

Reply:
The question is what happens with the storage nodes that are not upgrading? The satellite is pining the nodes once per hour. That is a kademlia ping. -> Upgraded satellite will be unable to talk to not upgraded storage nodes. -> Backwards incompatible.

Reply:
Why wouldn't it be able to talk to them? We have a method to fetch peer identity from a node without having a special endpoint.

Reply:
Because from the satellite point of view the nodes are offline because they can't be pinged.

Reply:
we should still have a ping method.

Reply:
Why do we need a ping? When we can establish a grpc connection that should be sufficient to get the same information.

Reply:
littleskunk and i realized on a call today that if we keep just the overlay.Ping rpc method, it will be much easier for a new satellite to work with old and new storage nodes.

Reply:
If a storage node is offline we will not select it for uploads and downloads. If the satellite is updated but the storage nodes are not updated and the satellite can't ping the storage nodes, all downloads should fail.

Reply:
what's the plan here?